spectral_cube/analysis_utilities.py

```diff
@@ -0,0 +1,310 @@
+import numpy as np
+
+from astropy import units as u
+from astropy.extern.six.moves import zip
+from astropy.extern.six.moves import range as xrange
+from astropy import log
+from astropy.wcs import WCS
+
+from .cube_utils import _map_context
+from .lower_dimensional_structures import OneDSpectrum
+
+
+def fourier_shift(x, shift, axis=0, add_pad=False, pad_size=None):
+    '''
+    Shift a spectrum in the Fourier plane.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Array to be shifted
+    shift : int or float
+        Number of pixels to shift.
+    axis : int, optional
+        Axis to shift along.
+    pad_size : int, optional
+        Pad the array before shifting.
+
+    Returns
+    -------
+    x2 : np.ndarray
+        Shifted array.
+
+    '''
+    mask = ~np.isfinite(x)
+    nonan = x.copy()
+
+    shift_mask = False
+    if not mask.all():
+        nonan[mask] = 0.0
+        shift_mask = True
+
+    # Optionally pad the edges
+    if add_pad:
+        if pad_size is None:
+            # Pad by the size of the shift
+            pad = np.ceil(shift).astype(int)
+
+            # Determine edge to pad whether it is a positive or negative shift
+            pad_size = (pad, 0) if shift > 0 else (0, pad)
+        else:
+            assert len(pad_size)
+
+        pad_nonan = np.pad(nonan, pad_size, mode='constant',
+                           constant_values=(0))
+        if shift_mask:
+            pad_mask = np.pad(mask, pad_size, mode='constant',
+                              constant_values=(0))
+    else:
+        pad_nonan = nonan
+        pad_mask = mask
+
+    nonan_shift = _fourier_shifter(pad_nonan, shift, axis)
+    if shift_mask:
+        mask_shift = _fourier_shifter(pad_mask, shift, axis) > 0.5
+        nonan_shift[mask_shift] = np.NaN
+
+    return nonan_shift
+
+
+def _padder(x, pad_size, shift):
+    '''
+    Pad x with zeros.
+    '''
+    return np.pad(x, pad_size, mode='constant',
+                  constant_values=(0))
+
+
+def _fourier_shifter(x, shift, axis):
+    '''
+    Helper function for `~fourier_shift`.
+    '''
+    ftx = np.fft.fft(x, axis=axis)
+    m = np.fft.fftfreq(x.shape[axis])
+    m_shape = [1] * len(x.shape)
+    m_shape[axis] = m.shape[0]
+    m = m.reshape(m_shape)
```

Comment:
I believe this could be rephrased as:
```
slices = [slice(None) if ii == axis else None for ii in range(x.ndim)]
m = m[slices]
```
I'm not sure that's actually any clearer...